src/Bicep.Core/Semantics/Namespaces/SystemNamespaceType.cs

```diff
@@ -281,7 +283,16 @@ private static IEnumerable<FunctionOverload> GetSystemOverloads(IFeatureProvider
 
             yield return new FunctionOverloadBuilder("first")
                 // TODO even with non-literal types, some type arithmetic could be performed
-                .WithReturnResultBuilder(TryDeriveLiteralReturnType("first", LanguageConstants.Any), LanguageConstants.Any)
+                .WithReturnResultBuilder((binder, fileResolver, diagnostics, arguments, argumentTypes) =>
+                {
+                    return argumentTypes[0] switch
+                    {
+                        TypedArrayType typedArrayType => new(typedArrayType.Item.Type),
```

Comment:
I would just use `ArrayType` as a matcher for this branch. `TypedArrayType` and `TupleType` are specializations of `ArrayType`, but both `ArrayType` and `TypedArrayType` will use `.Item.Type`. The `TupleType` matcher would need to come first, since `TupleType` is a subclass of `ArrayType`.

I also think that for now, `ArrayType`s will need to use a return type of `TypeHelper.CreateTypeUnion(LangaugeConstants.Null, arrayType.Item.Type)` (instead of just `arrayType.Item.Type`) since there's no way to tell at compile time whether the array is empty.

Reply:
Regarding returning the union-type for `ArrayType`, seems it's having trouble returning the custom type:
```
...
ArrayType arrayType => TypeHelper.CreateTypeUnion(LanguageConstants.Null, arrayType.Item.Type),
...
```
![image](https://user-images.githubusercontent.com/16405269/213054038-6a3befed-97cf-4f0d-a092-de40fe4a7cc8.png)

Reply:
From the error message, you can see that the custom type is being returned correctly, but autocomplete doesn't work on nullable types. That should be addressed in 0.14 (via https://github.com/Azure/bicep/pull/9368 and/or https://github.com/Azure/bicep/issues/9565).

The alternative is to just let the property access fail at runtime if the array passed to `first` is empty.